comictaggerlib/ui/talkeruigenerator.py

```diff
@@ -0,0 +1,239 @@
+            value_type = type(value)
+            try:
+                if value_type is str:
+                    widget.setText(value)
+                if value_type is int or value_type is float:
+                    widget.setValue(value)
+                if value_type is bool:
+                    widget.setChecked(value)
+            except Exception:
+                logger.debug("Failed to set value of %s", name)
+
+
+def form_settings_to_config(sources: dict[str, QtWidgets.QWidget], config: settngs.Config[settngs.Namespace]) -> None:
+    # Source combo box value
+    config[0].talker_source = sources["talker_source"].currentData()
+
+    for tab in sources["tabs"].items():
+        for name, widget in tab[1]["widgets"].items():
+            widget_value = None
+            if isinstance(widget, (QtWidgets.QSpinBox, QtWidgets.QDoubleSpinBox)):
+                widget_value = widget.value()
+            elif isinstance(widget, QtWidgets.QLineEdit):
+                widget_value = widget.text().strip()
+            elif isinstance(widget, QtWidgets.QCheckBox):
+                widget_value = widget.isChecked()
+
+            setattr(config[0], name, widget_value)
+
+
+def generate_source_option_tabs(
+    comic_talker_tab: QtWidgets.QWidget,
+    config: settngs.Config[settngs.Namespace],
+    talkers: dict[str, ComicTalker],
+) -> dict[str, QtWidgets.QWidget]:
+    """
+    Generate GUI tabs and settings for talkers
+    """
+
+    # Store all widgets as to allow easier access to their values vs. using findChildren etc. on the tab widget
+    sources: dict = {"tabs": {}}
+
+    # Tab comes with a QVBoxLayout
+    comic_talker_tab_layout = comic_talker_tab.layout()
+
+    talker_layout = QtWidgets.QGridLayout()
+    lbl_info = QtWidgets.QLabel("Information Source:")
+    cbx_info = QtWidgets.QComboBox()
+    line = QtWidgets.QFrame()
+    line.setFrameShape(QtWidgets.QFrame.HLine)
+    line.setFrameShadow(QtWidgets.QFrame.Sunken)
+    talker_tabs = QtWidgets.QTabWidget()
+
+    talker_layout.addWidget(lbl_info, 0, 0, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Maximum)
+    talker_layout.addWidget(cbx_info, 0, 1, QtWidgets.QSizePolicy.Minimum, QtWidgets.QSizePolicy.Maximum)
+    talker_layout.addWidget(line, 1, 0, 1, -1)
+    talker_layout.addWidget(talker_tabs, 2, 0, 1, -1)
+
+    comic_talker_tab_layout.addLayout(talker_layout)
+
+    # Add cbx_info combobox to sources for getting and setting talker
+    sources["talker_source"] = cbx_info
+
+    # Add source sub tabs to Comic Sources tab
+    for talker_id, talker_obj in talkers.items():
+        # Add source to general tab dropdown list
+        cbx_info.addItem(talker_obj.name, talker_id)
+
+        tab_name = talker_id
+        sources["tabs"][tab_name] = {"tab": QtWidgets.QWidget(), "widgets": {}}
+        layout_grid = QtWidgets.QGridLayout()
+
+        for option in config[1][f"talker_{talker_id}"][1].values():
+            current_widget = None
```

Comment:
> It's not so much about making it easier to generate, more that most APIs don't require a key so there will only be a URL to test. On that note, it looks like you've made every talker have an API key entry?

I haven't checked all of the different comic/manga websites for what their API requires, but in general most websites require that you register an account to make API queries.

Now that the key and url settings always exist, an arbitrary extension check (`.endswith("key")`) is not needed.
All of them are `{id}_key` and `{id}_url` so this is known before hand and can be added with only one button, how it was before. Before looping over the settings definition `config[1]` the key and url text boxes can be put in with a single button to test the API and then skip those two names when generating the rest of the settings.
If they don't wan't the settings to show up at all they can use
```python
        parser.add_setting(f"--{self.id}-key", file=False, cmdline=False)
        parser.add_setting(f"--{self.id}-url", file=False, cmdline=False)
```
and it will not show up in any settings (if there is already one saved in the settings file the value will stay but will be unchangeable from the CLI or the settings window) , the only place it will show up is on the definitions

And this check needs to be added, if a talker doesn't plan on saving the setting then it doesn't need to be in the GUI.

```suggestion
        for option in config[1][f"talker_{talker_id}"][1].values():
            if not option.file:
                continue
            current_widget = None
```

Reply:
An additional small gripe is that the URL and key are always on top in the GUI now. Before the order was dictated by the order within the talker. It's a personal preference but I think it looks better when the check boxes are first. I could order them with the generation but that takes the order out of the talkers hands as well. Very minor but seeing a text field followed by check boxes and then possibly another text field will hurt a little :)

Now you've added a `display_name` to settngs, is an `order` field worth any consideration?

Reply:
Another reason for separate test buttons and functions is not knowing which one has failed. More logic could be placed in the test on the talker and return a message rather than True/False.

If going with only one button, it would have to go on the URL line (because key is optional with some talkers). It would then make sense to have a tool tip explaining it will test both.

Does all that check out?

Reply:
I don't have an issue either way they just need to be next to each other you can easily just add it after you add everything else.

No, not adding an order field, plugins can manage order by what order they call `add_setting`.

That's something that can be worked on later, currently that's all it gives for ComicVine, it will work for any others for now, and they can always write to the log.

I'm sure someone will decide the url is optional but not the key.
```python
def add_api_widgets(config: settngs.Config[settngs.Namespace], layout: QtWidgets.QGridLayout, talker: ComicTalker):
    # *args enforces keyword arguments and allows position arguments to be ignored
    def call_check_api(*args, le_url: QtWidgets.QLineEdit, le_key: QtWidgets.QLineEdit, talker: ComicTalker):
        url = ""
        key = ""
        if le_key is not None:
            key = le_key.text().strip()
        if le_url is not None:
            url = le_url.text().strip()
        if talker.check_api_key(key, url):
            QtWidgets.QMessageBox.information(None, "API Test", "Key is valid!")
        else:
            QtWidgets.QMessageBox.warning(None, "API Test", "Key is NOT valid!")

    # get the actual config objects in case they have overwritten the default
    talker_key = config[1][f"talker_{talker.id}"][1][f"{talker.id}_key"]
    talker_url = config[1][f"talker_{talker.id}"][1][f"{talker.id}_url"]
    btn_test_row = None
    le_key = None
    le_url = None

    # only file settings are saved
    if talker_key.file:
        # record the current row so we know where to add the button
        btn_test_row = layout.rowCount()
        le_key = generate_textbox(talker_key, layout)

    # only file settings are saved
    if talker_url.file:
        # record the current row so we know where to add the button
        # We overwrite so that the default will be next to the url text box
        btn_test_row = layout.rowCount()
        le_url = generate_textbox(talker_url, layout)

    # The button row was recorded so we add it
    if btn_test_row is not None:
        btn = QtWidgets.QPushButton("Test API")
        layout.addWidget(btn, btn_test_row, 2)
        # partial is used as connect will pass in event information
        btn.clicked.connect(partial(call_check_api, le_url=le_url, le_key=le_key, talker=talker))
```

Then you call `add_api_widgets(config, layout_grid, talker_obj)` either before or after the rest of the widgets and you add this to the beginning of the widget generation loop:
```python
            if option.dest in (f"{talker_id}_url", f"{talker_id}_key"):
                continue
```

Reply:
I'd already started on something similar and so left the test API string return stuff in. I can take it out if desired?

Reply:
Make it return a `tuple[str, bool]`